controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,895 @@
+		upstreamNP.Mode = string(np.Mode)
+		upstreamNP.OsType = string(np.OsType)
+		upstreamNP.OrchestratorVersion = np.OrchestratorVersion
+		upstreamNP.AvailabilityZones = np.AvailabilityZones
+		if np.EnableAutoScaling != nil {
+			upstreamNP.EnableAutoScaling = np.EnableAutoScaling
+			upstreamNP.MaxCount = np.MaxCount
+			upstreamNP.MinCount = np.MinCount
+		}
+		upstreamSpec.NodePools = append(upstreamSpec.NodePools, upstreamNP)
+	}
+
+	// set network configuration
+	networkProfile := clusterState.NetworkProfile
+	if networkProfile != nil {
+		upstreamSpec.NetworkPlugin = string(networkProfile.NetworkPlugin)
+		upstreamSpec.NetworkDNSServiceIP = to.String(networkProfile.DNSServiceIP)
+		upstreamSpec.NetworkDockerBridgeCIDR = to.String(networkProfile.DockerBridgeCidr)
+		upstreamSpec.NetworkServiceCIDR = to.String(networkProfile.ServiceCidr)
+		upstreamSpec.NetworkPolicy = string(networkProfile.NetworkPolicy)
+		upstreamSpec.NetworkPodCIDR = to.String(networkProfile.PodCidr)
+		upstreamSpec.LoadBalancerSKU = string(networkProfile.LoadBalancerSku)
+	}
+
+	// set linux account profile
+	linuxProfile := clusterState.LinuxProfile
+	if linuxProfile != nil {
+		upstreamSpec.LinuxAdminUsername = to.String(linuxProfile.AdminUsername)
+		sshKeys := *linuxProfile.SSH.PublicKeys
+		upstreamSpec.LinuxSSHPublicKeyContents = to.String(sshKeys[0].KeyData)
+	}
+
+	// set API server access profile
+	upstreamSpec.PrivateCluster = to.BoolPtr(false)
+	if clusterState.APIServerAccessProfile != nil {
+		if clusterState.APIServerAccessProfile.EnablePrivateCluster != nil {
+			upstreamSpec.PrivateCluster = clusterState.APIServerAccessProfile.EnablePrivateCluster
+		}
+		if clusterState.APIServerAccessProfile.AuthorizedIPRanges != nil {
+			upstreamSpec.AuthorizedIPRanges = clusterState.APIServerAccessProfile.AuthorizedIPRanges
+		}
+	}
+
+	return upstreamSpec, err
+}
+
+// updateUpstreamClusterState compares the upstream spec with the config spec, then updates the upstream AKS cluster to
+// match the config spec. Function returns after a update is finished.
+func updateUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.SecretCache, spec aksv1.AKSClusterConfigSpec, upstreamSpec *aksv1.AKSClusterConfigSpec) (*aksv1.AKSClusterConfigSpec, error) {
+	// check kubernetes version for update
+	if spec.Tags == nil {
+		spec.Tags = map[string]string{}
+	}
+	spec.Tags["displayName"] = spec.DisplayName
+	if spec.Imported || reflect.DeepEqual(spec.Tags, upstreamSpec.Tags) &&
+		spec.NodePools != nil && reflect.DeepEqual(spec.NodePools, upstreamSpec.NodePools) {
+		return nil, nil
+	}
+
+	azureAuthorizer, err := newClientAuthorizer(secretsCache, spec)
+	if err != nil {
+		return nil, err
+	}
+	resourceGroupsClient, err := newResourceGroupsClient(azureAuthorizer, spec)
+	if err != nil {
+		return nil, err
+	}
+
+	if !existsResourceGroup(ctx, resourceGroupsClient, spec.ResourceGroup) {
+		logrus.Infof("Resource group [%s] does not exist, creating", spec.ResourceGroup)
```

Comment:
Not sure I follow this one - would we ever need to create a new resource group for an upstream (updated) cluster? Wouldn't that resource group already exist?

Reply:
It can be scenario, when user will remove AKS and reource group on the cloud provider site, so we have to be ready to recreate full cluster.